src/core/engines/cross-margin/CrossMarginMath.sol

```diff
@@ -0,0 +1,564 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.0;
+
+import {SafeCast} from "openzeppelin/utils/math/SafeCast.sol";
+
+import {IGrappa} from "../../../interfaces/IGrappa.sol";
+import {IOracle} from "../../../interfaces/IOracle.sol";
+
+import {AccountUtil} from "../../../libraries/AccountUtil.sol";
+import {ArrayUtil} from "../../../libraries/ArrayUtil.sol";
+import {NumberUtil} from "../../../libraries/NumberUtil.sol";
+import {ProductIdUtil} from "../../../libraries/ProductIdUtil.sol";
+import {TokenIdUtil} from "../../../libraries/TokenIdUtil.sol";
+
+import "../../../config/constants.sol";
+import "../../../config/enums.sol";
+import "../../../config/types.sol";
+import "../../../config/errors.sol";
+
+/**
+ * @title   CrossMarginMath
+ * @notice  this library is in charge of calculating the min collateral for a given cross margin account
+ * @dev     deployed as a separate contract to save space
+ */
+library CrossMarginMath {
+    using AccountUtil for Balance[];
+    using AccountUtil for CrossMarginDetail[];
+    using AccountUtil for Position[];
+    using AccountUtil for PositionOptim[];
+    using AccountUtil for SBalance[];
+    using ArrayUtil for uint256[];
+    using ArrayUtil for int256[];
+    using SafeCast for int256;
+    using SafeCast for uint256;
+    using TokenIdUtil for uint256;
+
+    error CM_InvalidPutLengths();
+
+    error CM_InvalidCallLengths();
+
+    error CM_InvalidPutWeight();
+
+    error CM_InvalidCallWeight();
+
+    error CM_InvalidPoints();
+
+    error CM_InvalidLeftPointLength();
+
+    error CM_InvalidRightPointLength();
+
+    /*///////////////////////////////////////////////////////////////
+                         Portfolio Margin Requirements
+    //////////////////////////////////////////////////////////////*/
+
+    function getMinCollateralForAccount(IGrappa grappa, CrossMarginAccount calldata account)
```

Comment:
I do think we should change the name here: because we're not returning the minimum collateral, but instead substracting collateral by required amount, and return negative number if it's underwater

Reply:
not anymore - https://github.com/antoncoding/grappa/blob/refactor/cm_math_ints/src/core/engines/cross-margin/CrossMarginMath.sol#L63

Reply:
once i get all the documentation in there, I am going to create a PR.

Reply:
sweet, you should avoid complex wording like `numeraireNeeded` tho 🤣 , just call it `collateralNeeded `